core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt

```diff
@@ -0,0 +1,174 @@
+      return when {
+        storage == null -> defaultStorage().also(::putPrefStorage)
+        !File(storage).exists() -> defaultStorage()
+        else -> storage
+      }
+    }
+
+  private fun defaultStorage(): String =
+    getExternalFilesDirs(context, null)[0]?.path
+      ?: context.filesDir.path  // a workaround for emulators
+
+  fun getPrefStorageTitle(defaultTitle: String): String =
+    sharedPreferences.getString(PREF_STORAGE_TITLE, defaultTitle) ?: defaultTitle
+
+  fun putPrefLanguage(language: String) =
+    sharedPreferences.edit { putString(PREF_LANG, language) }
+
+  fun putPrefIsFirstRun(isFirstRun: Boolean) =
+    sharedPreferences.edit { putBoolean(PREF_IS_FIRST_RUN, isFirstRun) }
+
+  fun putPrefWifiOnly(wifiOnly: Boolean) =
+    sharedPreferences.edit { putBoolean(PREF_WIFI_ONLY, wifiOnly) }
+
+  fun putPrefStorageTitle(storageTitle: String) =
+    sharedPreferences.edit { putString(PREF_STORAGE_TITLE, storageTitle) }
+
+  fun putPrefStorage(storage: String) {
+    sharedPreferences.edit { putString(PREF_STORAGE, storage) }
+    _prefStorages.onNext(storage)
+  }
+
+  fun putPrefFullScreen(fullScreen: Boolean) =
+    sharedPreferences.edit { putBoolean(PREF_FULLSCREEN, fullScreen) }
+
+  fun putPrefExternalLinkPopup(externalLinkPopup: Boolean) =
+    sharedPreferences.edit { putBoolean(PREF_EXTERNAL_LINK_POPUP, externalLinkPopup) }
+
+  fun showIntro(): Boolean = sharedPreferences.getBoolean(PREF_SHOW_INTRO, true)
+
+  fun setIntroShown() = sharedPreferences.edit { putBoolean(PREF_SHOW_INTRO, false) }
+
+  var showHistoryAllBooks: Boolean
+    get() = sharedPreferences.getBoolean(PREF_SHOW_HISTORY_ALL_BOOKS, true)
+    set(prefShowHistoryAllBooks) {
+      sharedPreferences.edit { putBoolean(PREF_SHOW_HISTORY_ALL_BOOKS, prefShowHistoryAllBooks) }
+    }
+
+  var showBookmarksAllBooks: Boolean
+    get() = sharedPreferences.getBoolean(PREF_SHOW_BOOKMARKS_ALL_BOOKS, true)
+    set(prefShowBookmarksFromCurrentBook) = sharedPreferences.edit {
+      putBoolean(PREF_SHOW_BOOKMARKS_ALL_BOOKS, prefShowBookmarksFromCurrentBook)
+    }
+
+  val nightMode: NightModeConfig.Mode
+    get() = from(
+      sharedPreferences.getString(PREF_NIGHT_MODE, null)?.toInt()
+        ?: AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
+    )
+
+  fun nightModes(): Flowable<NightModeConfig.Mode> = nightModes.startWith(nightMode)
+
+  fun updateNightMode() = nightModes.offer(nightMode)
+
+  var hostedBooks: Set<String>
+    get() = sharedPreferences.getStringSet(PREF_HOSTED_BOOKS, null)?.toHashSet() ?: HashSet()
+    set(hostedBooks) {
+      sharedPreferences.edit { putStringSet(PREF_HOSTED_BOOKS, hostedBooks) }
+    }
+
+  var textZoom: Int
+    get() = sharedPreferences.getInt(TEXT_ZOOM, DEFAULT_ZOOM)
+    set(textZoom) {
+      sharedPreferences.edit { putInt(TEXT_ZOOM, textZoom) }
+      textZooms.offer(textZoom)
+    }
+
+  fun getTextZooms(): Flowable<Int> = textZooms.startWith(textZoom)
```

Comment:
alt+enter to convert this into a val with a getter after reworking `textZooms` to be a private backing field, for consistency

Reply:
I have two classes scheduled in line from 4pm(after 5 mins from now), i'll try finishing it right now, otherwise afterwards